Added remove_row() and remove_col()

djkato opened the pull request:
Adding functions `remove_col(index: usize)` and `remove_row(index: usize)`.

becheran commented on src/lib.rs:
```diff
@@ -1613,4 +1660,21 @@ r#"[
         assert_eq!(sum_by_col, vec![1+4, 2+5, 3+6])
     }
+        #[test]
+    fn remove_col_empty() {
```
Remove empty from fn name

becheran commented on src/lib.rs:
```diff
@@ -1613,4 +1660,21 @@ r#"[
+        assert_eq![grid.remove_col(0), None];
+    }
+    #[test]
+    fn remove_row_empty() {
```
Same

becheran commented on src/lib.rs:
```diff
@@ -1613,4 +1660,21 @@ r#"[
         let sum_by_col: Vec<u8> = grid.iter_cols().map(|col| col.sum()).collect();
         assert_eq!(sum_by_col, vec![1+4, 2+5, 3+6])
     }
+        #[test]
```
Thx for adding tests. Could you please also add tests for out of bounds check? For example empty grid or idx greater size?

becheran commented on src/lib.rs:
````diff
@@ -559,6 +559,29 @@ impl<T> Grid<T> {
+    /// assert_eq![grid.remove_row(0), None];
+    /// ```
+    pub fn remove_row(&mut self, row_index: usize) -> Option<Vec<T>> {
+        if self.cols == 0 || self.rows == 0 || row_index > self.rows {
````
Shouldn't it be >= ?

becheran commented on src/lib.rs:
````diff
@@ -589,6 +612,30 @@ impl<T> Grid<T> {
+    /// assert_eq![grid.remove_col(0), None];
+    /// ```
+    pub fn remove_col(&mut self, col_index: usize) -> Option<Vec<T>> {
+        if self.cols == 0 || self.rows == 0 || col_index > self.cols {
````
>= ?

becheran reviewed (changes requested):
Thx for the pr. Some remarks

djkato commented:
Thanks for the >=, didn't catch that. Renamed existing tests and added out of bound tests. 

I'd recommend running a formatter on the project. the standard rustfmt makes a huge amount of changes and one can tell it's not  been formatted in a while :).

becheran commented on src/lib.rs:
```diff
@@ -589,6 +612,30 @@ impl<T> Grid<T> {
+        }
+        let mut residue: Vec<T> = vec![];
+        for i in 0..self.rows {
+			residue.push(self.data.remove(i * self.cols + col_index - i));
```
Another thing that just came into my mind now is that this is quite inefficient. Every iteration the remaining vector would be reordered for every remove call.

djkato replied:
I played around with optimizing this a bit but sorry I'm relatively new to rust and this is my first github pull request :).
I don't see how every iteration reorders anything, as `push()` appends to the end of the residue vector and `remove()` keeps the order of the original vector. I can't really talk about speed of `remove()` though.
Googling around I found that a faster method to remove the elements would be either `swap_remove()` which reorders the elements, or `retain()`, which would be faster but then I loose the option to return the `Some(residue)`, as:
```rs
let mut i  = 0;
let total_row = self.rows.clone();
self.data.retain(|element: &T| {
    i += 1;
    if ((i- 1)% total_row) == col_index  {
        // residue.push(element.to_owned()); cannot clone reference...
        println!("Removing at {:?}", i - 1);
        return false
    } else {
        return true
    }
});
```
Using a custom loop that splits the vector always at col_index seems more memory intensive as there are multiple new vectors allocated. While the speed of `remove()` isn't the best I don't see another way myself... If you have an idea I'd love to hear it!

becheran replied:
No worries.

The remove function is slow because one has to sort the elements after the removal one position to the left. Doing this in a loop we end up with a runtime complexity of _O(n^2)_. Swap remove is very fast because it is essentially just a swap of last and current element and then decrease the length by one _O(1)_. But, this won't help because it will mess up the order. I think the best possible result is only iterating over all elements once _O(n)_.

What I did in the pop column is using rotations and in the end decrease the size. Anyhow if in doubt about performance it is always good to proof improvements via benchmarks.

You can run the existing benchmark with _cargo bench_.

I can also do this if you want. I just don't know when I will find some time.

djkato replied:
Yeah I understand it's not optimal, I might try to solve it later when I have time. Maybe you'll even have time sooner than me :). I'm already using my version of it though cause I needed it locally. I Made this pull request cause any remove function is better than no function.

becheran replied:
I pushed a version with uses rotations instead of vec `remove()`. What do you think? The benchmark shows an increase from 120 ms to now only 90 us for a 1000*1000 grid.

djkato replied:
Not sure I fully understand what's going on there but if it improves performance then great 👌 
Thanks for finding time.

becheran reviewed (changes requested):
The removal of cols can be more efficient. Take a look at the pop_col fn. I did some work there to remove the last col more efficient.
